libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+      break;
+    }
+    __storage_.__large_.__status_ = _NotEngaged;
+  }
+
+public:
+  using result_type = _ReturnT;
+
+  // [func.wrap.move.ctor]
+  move_only_function() noexcept = default;
+  _LIBCPP_HIDE_FROM_ABI move_only_function(nullptr_t) noexcept : __storage_({}) {}
+  _LIBCPP_HIDE_FROM_ABI move_only_function(move_only_function&& __other) noexcept : __storage_(__other.__storage_) {
+    __other.__storage_ = {};
+  }
+
+  template <class _Func>
+    requires(!is_same_v<remove_cvref_t<_Func>, move_only_function> && !__is_inplace_type<_Func>::value &&
+             __is_callable_from<_Func>)
+  _LIBCPP_HIDE_FROM_ABI move_only_function(_Func&& __func) noexcept {
+    using _FuncWraps = __function_wrappers<_Func>;
+    using _UnRefFunc = remove_reference_t<_Func>;
+
+    if constexpr ((is_pointer_v<_UnRefFunc> && is_function_v<remove_pointer_t<_UnRefFunc>>) ||
+                  is_member_function_pointer_v<_UnRefFunc>) {
+      if (__func == nullptr) {
+        __storage_ = {};
+        return;
+      } else {
+        auto& __data   = __storage_.__trivially_destructible_;
+        __data.__call_ = std::bit_cast<void*>(&_FuncWraps::__call);
+        std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__data_), std::forward<_Func>(__func));
+        __data.__status_ = _Store::_Status::_TriviallyDestructible;
+      }
+    } else if constexpr (__is_move_only_function<remove_cvref_t<_Func>>::value) {
+      if (!__func) {
+        __storage_ = {};
+      } else {
+        auto& __data      = __storage_.__large_;
+        __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+        __data.__destroy_ = &_FuncWraps::__destroy;
+        __data.__heap_    = static_cast<std::byte*>(::operator new(sizeof(_Func), std::align_val_t(alignof(_Func))));
+        std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__heap_), std::forward<_Func>(__func));
+        __data.__status_ = _Store::_Status::_Heap;
+      }
+    } else {
+      __construct<_Func>(std::forward<_Func>(__func));
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI _Store::_Status __get_status() noexcept { return __storage_.__large_.__status_; }
+
+  template <class _Func, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(in_place_type_t<_Func>, _Args&&... __args) {
+    static_assert(is_same_v<decay_t<_Func>, _Func>);
+    __construct<_Func>(std::forward<_Args>(__args)...);
+  }
+
+  template <class _Func, class _InitListType, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, initializer_list<_InitListType>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(
+      in_place_type_t<_Func>, initializer_list<_InitListType> __il, _Args&&... __args) {
+    __construct<_Func>(__il, std::forward<_Args>(__args)...);
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
```

Comment:
Good catch! FWIW, I would say yes, and also that this deserves an LWG issue. Assignment operators (along with ctors and `swap`) are generally marked noexcept where possible, _especially_ (as here) when the `noexcept` can be unconditional. This is [P0884R0](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0884r0.pdf) rule C.

Reply:
LWG points out that `std::function`'s move-assignment operator is specified as non-`noexcept` because of the rationale given in [P0771R1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0771r1.pdf):

>  if allocators are added back to the `std::function` interface, the move-assignment operator behaves as a copy assignment under certain conditions, making the noexcept guarantee impossible.

However, all three vendors do mark `std::function`'s move-assignment operator as `noexcept` in practice. So we should do the same thing here: we _should_ mark it `noexcept` for consistency with our own `std::function`; but it is _not_ worthy of an LWG issue, because the wording is consistent with the intent.

Reply:
Thanks for looking into that! Removed the comment.